src/runtime/entity.ts

```diff
@@ -385,12 +385,20 @@ function getInternals(entity): EntityInternals {
 }
 
 function sanitizeAndApply(target: Entity, data: EntityRawData, schema: Schema, context: ChannelConstructor) {
+  const temp = {...target};
   for (const [name, value] of Object.entries(data)) {
     const sanitizedValue = Entity.sanitizeEntry(schema.fields[name], value, name, context);
     Entity.validateFieldAndTypes(name, sanitizedValue, schema);
-    target[name] = sanitizedValue;
+    temp[name] = sanitizedValue;
   }
   if (Flags.enforceRefinements) {
-    Refinement.refineData(target, schema);
+    const exception = Refinement.refineData(temp, schema);
+    if (exception) {
+      context.reportExceptionInHost(exception);
+      return;
+    }
+  }
+  for (const [name, value] of Object.entries(temp)) {
```

Comment:
I think it's worth adding a comment to explain why this is being done (its currently in danger of being 'cleaned up').